checks.d/openstack.py

```diff
@@ -0,0 +1,421 @@
+
+        hypervisor_ids = []
+        try:
+            resp = self._make_request_with_auth_fallback(url, headers)
+            hv_list = resp.json()
+            for hv in hv_list['hypervisors']:
+                hypervisor_ids.append(hv['id'])
+        except Exception as e:
+            self.warning('Unable to get the list of all hypervisors: {0}'.format(str(e)))
+
+        return hypervisor_ids
+
+    def get_uptime_for_single_hypervisor(self, hyp_id):
+        url = '{0}/os-hypervisors/{1}/uptime'.format(self._nova_url, hyp_id)
+        headers = {'X-Auth-Token': self._auth_token}
+
+        resp = self._make_request_with_auth_fallback(url, headers)
+        uptime = resp.json()['hypervisor']['uptime']
+        return self._parse_uptime_string(uptime)
+
+    def get_stats_for_single_hypervisor(self, hyp_id):
+        url = '{0}/os-hypervisors/{1}'.format(self._nova_url, hyp_id)
+        headers = {'X-Auth-Token': self._auth_token}
+        resp = self._make_request_with_auth_fallback(url, headers)
+        hyp = resp.json()['hypervisor']
+
+        hyp_state = hyp.get('state', None)
+        if hyp_state is None:
+            try:
+                # Fall back for pre Nova v2.1 to the uptime response
+                uptime = self.get_uptime_for_single_hypervisor(hyp_id)
+                if uptime.get('uptime_sec', 0) > 0:
+                    hyp_state = self.HYPERVISOR_STATE_UP
+                else:
+                    hyp_state = self.HYPERVISOR_STATE_DOWN
+            except:
+                # This creates the AgentCheck.UNKNOWN state
+                pass
+
+        service_check_tags = [
+            'hypervisor:{0}'.format(hyp['hypervisor_hostname']),
+            'hypervisor_id:{0}'.format(hyp['id']),
+            'virt_type:{0}'.format(hyp['hypervisor_type'])
+        ]
+
+        if hyp_state is None:
+            self.service_check(self.HYPERVISOR_SERVICE_CHECK_NAME, AgentCheck.UNKNOWN,
+                               tags=service_check_tags)
+        elif hyp_state != self.HYPERVISOR_STATE_UP:
+            self.service_check(self.HYPERVISOR_SERVICE_CHECK_NAME, AgentCheck.CRITICAL,
+                               tags=service_check_tags)
+        else:
+            self.service_check(self.HYPERVISOR_SERVICE_CHECK_NAME, AgentCheck.OK,
+                               tags=service_check_tags)
+
+        return hyp
+
+    def get_server_stats(self):
+        if self.init_config.get('check_all_servers', False):
+            server_ids = self.get_all_server_ids()
+        else:
+            server_ids = self.init_config.get('server_ids', [])
+
+        if not server_ids:
+            self.warning("Your check is not configured to monitor any servers.\n" +
+                         "Please list `server_ids` under your init_config in openstack.yaml")
+
+        for sid in server_ids:
+            self.get_stats_for_single_server(sid)
+
+    def get_all_server_ids(self):
+        url = '{0}/servers'.format(self._nova_url)
```

Comment:
Needs to have have the option to have `?all_tenants=1` in the URL or else it will only get all servers for the current project/tenant. Also, FYI, if the auth scope is not 'admin' then it will still limit the results to the current tenant (no errors).